add granting privileges for a read-only user for doing remote pg_dump

milberg1 opened the pull request:
pull request for add_readonly_user branch

eshellman commented:
is there any reason not to treat this as a migration? i.e. can the backupuser be a permanent part of the schema the way "gutenberg" is?

eshellman commented:
suggested test:
1. apply script
2. pg_dump -U gutenberg -s #dump schema
3. pg_dump -U backupuser -d # full backup
4. pg_restore # what user?
5. pg_dump -U gutenberg -s #restored schema should be same as before

milberg1 commented:
I think being a permanent part of the schema would be good, but wasn't sure
how that would fit with the existing gutenberg_public.sql -- would it need
to have a CREATE USER statement for backupuser?  I avoided that to keep the
password out of this.

For testing:

   - Would step 2 include "-s --schema=public" ?  (public is the only
   schema the grants would give access to)
   - A least-impact test would be to restore to a DB with a different name,
   but then backupuser wouldn't have access to dump the restored one.  I can
   imagine 2 options: either restore to different DB, then have postgres user
   dump new DB and compare the 2 dumps; or could dump gutenberg_dev, drop
   gutenberg_dev, restore gutenberg_dev, and again dump gutenberg_dev and
   compare the 2 dumps.  This would impact the dev site.  If we ended up with
   a broken gutenberg_dev, could restore it from gutenberg.  Were you thinking
   of one of these options, or something else?
   - I think pg_restore should be done locally as user postgres.




On Tue, May 25, 2021 at 12:05 PM Eric Hellman ***@***.***>
wrote:

> suggested test:
>
>    1. apply script
>    2. pg_dump -U backupuser -s #dump schema
>    3. pg_dump -U backupuser -d # full backup
>    4. pg_restore # what user?
>    5. pg_dump -U backupuser -s #restored schema should be same as before
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gutenbergtools/pgdb/pull/3#issuecomment-847999257>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AS2QJV3GJ6BFJLJS5WMZGN3TPPDFHANCNFSM45PVFF3A>
> .
>

eshellman commented:
1. I think if you don't have the user, the pg restore will ignore it.
2. -s --schema=public is fine
3. I think it's ok to mess with the dev site
4. yes, restore as gutenberg seems necessary.

eshellman commented:
I haven't committed the schemas for other than local; I will add them.
Can the script be broken down into one script per schema? Because since we should probably remove those schemas/tables completely, we won't have to modify the script for public.

milberg1 commented:
Thanks for looking at this.  Splitting into one script per schema sounds
good.  I think the 3 GRANT CONNECT statements would go in the script for
public.  Would you prefer to make this update or for me to?


On Tue, Jun 15, 2021 at 9:08 AM Eric Hellman ***@***.***>
wrote:

> I haven't committed the schemas for other than local; I will add them.
> Can the script be broken down into one script per schema? Because since we
> should probably remove those schemas/tables completely, we won't have to
> modify the script for public.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gutenbergtools/pgdb/pull/3#issuecomment-861482951>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AS2QJVZJOZ4RUVIMFWENGSLTS5GFNANCNFSM45PVFF3A>
> .
>

eshellman commented:
I've added all the schemas from prod; it looks like the script has already been applied?
I've also split the update script. I've assumed the database-wide preamble can be omitted from 03, 04, 05, and 06

@milberg1 please look at the diff for the public schema. I'll add a note in a few lines I don't know why they're there

eshellman commented on gutenberg_public.sql:
```diff
@@ -2859,171 +2859,435 @@ ALTER TABLE ONLY public.tweets
 --
 
 REVOKE ALL ON SEQUENCE public.aliases_pk_seq FROM gutenberg;
+GRANT USAGE ON SEQUENCE public.aliases_pk_seq TO gutenberg;
```
Why would there have been a revoke?

eshellman commented on gutenberg_public.sql:
```diff
@@ -2859,171 +2859,435 @@ ALTER TABLE ONLY public.tweets
 -- Name: SEQUENCE authors_pk_seq; Type: ACL; Schema: public; Owner: gutenberg
 --
 
 REVOKE ALL ON SEQUENCE public.authors_pk_seq FROM gutenberg;
```
same thing occurs in a few places.

eshellman commented on gutenberg_public.sql:
```diff
@@ -2859,171 +2859,435 @@ ALTER TABLE ONLY public.tweets
+-- Name: DEFAULT PRIVILEGES FOR SEQUENCES; Type: DEFAULT ACL; Schema: public; Owner: postgres
+--
+
+ALTER DEFAULT PRIVILEGES FOR ROLE postgres IN SCHEMA public REVOKE ALL ON SEQUENCES  FROM postgres;
```
here's another revoke.

milberg1 commented:
I got your updates, thanks!  Yes, I did the grants for backupuser already.
I agree that the preamble can be omitted from 03, 04, 05, and 06.


On Wed, Jun 16, 2021 at 2:50 PM Eric Hellman ***@***.***>
wrote:

> I've added all the schemas from prod; it looks like the script has already
> been applied?
> I've also split the update script. I've assumed the database-wide preamble
> can be omitted from 03, 04, 05, and 06
>
> @milberg1 <https://github.com/milberg1> please look at the diff for the
> public schema. I'll add a note in a few lines I don't know why they're there
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gutenbergtools/pgdb/pull/3#issuecomment-862626881>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AS2QJVZBXGO7ODJS2JZX57TTTDW7VANCNFSM45PVFF3A>
> .
>

eshellman commented:
@milberg1 what's the status of testing the restore?

milberg1 commented:
I had something unexpected come up that is almost resolved now.  Would it
be OK for me to test this afternoon?



On Thu, Jun 17, 2021 at 9:13 AM Eric Hellman ***@***.***>
wrote:

> @milberg1 <https://github.com/milberg1> what's the status of testing the
> restore?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gutenbergtools/pgdb/pull/3#issuecomment-863227982>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AS2QJV3R5543UUCF73VUF3LTTHYHPANCNFSM45PVFF3A>
> .
>
